python/ray/data/read_api.py

```diff
@@ -1608,6 +1608,13 @@ def read_tfrecords(
     )
     ds = read_datasource(datasource, parallelism=parallelism)
 
+    if fast_read:
+        from ray.data.datasource.tfrecords_datasource import (
+            _cast_large_list_to_list_transform,
+        )
+
+        ds = _cast_large_list_to_list_transform(ds)
+
     if fast_read_auto_infer_schema and fast_read and not tf_schema:
         from ray.data.datasource.tfrecords_datasource import _infer_schema_and_transform
 
```

Comment:
nit: we could combine this logic under the same if block for the fast path:
```suggestion
    if fast_read:
        from ray.data.datasource.tfrecords_datasource import (
            _cast_large_list_to_list_transform,
        )

        ds = _cast_large_list_to_list_transform(ds)

        if fast_read_auto_infer_schema and not tf_schema:
            from ray.data.datasource.tfrecords_datasource import _infer_schema_and_transform

```

Reply:
done in https://github.com/ray-project/ray/pull/42277/commits/681f7531621cf9a2cf1a8c2757b997182c644996